sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+package com.azure.android.core.http.responsepaging;
+
+import com.azure.android.core.http.SimpleCallback;
+import com.azure.android.core.util.paging.Page;
+
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import okhttp3.Response;
+
+/**
+ * Represents a collection of pages where page data and response for retrieving it can be returned asynchronously
+ * @param <T> type of the items in the page
+ * @param <P> Page of items
+ */
+public class AsyncPagedDataCollection<T, P extends Page<T>> {
+    private final AsyncPagedDataRetriever<T, P> pagedDataRetriever;
+    private LinkedHashMap<String, Map.Entry<Response, P>> pages = new LinkedHashMap<String, Map.Entry<Response, P>>();
+    private String firstPageId;
+
+    /**
+     * Construction requires an asynchronous paged data provider
+     * @param pagedDataRetriever an asynchronous page data provider
+     */
+    public AsyncPagedDataCollection(AsyncPagedDataRetriever<T, P> pagedDataRetriever) {
+        this.pagedDataRetriever = pagedDataRetriever;
+    }
+
+    private void cacheResponse(P page, Response response ){
+        pages.put(page.getPageId(), new Map.Entry<Response, P>() {
```

Comment:
Do we expect that the value of a given page will change for a given `Response`? I feel that we should use a different data structure to represent this pair, we could even have our own custom immutable `SimplePair`.

Reply:
Good idea. I didn't like Map.Entry at all but Pair is only available for SDK version 24 (26?) or later and we'd like to support 21